File: src/primitives/Address/toChecksummed.ts
import type { Address } from "./Address.js";
import * as Checksummed from "./ChecksumAddress.js";

/**
 * Convert Address to EIP-55 checksummed hex string
 *
 * @returns Checksummed hex string with mixed case
 *
 * @example
 * ```typescript
 * const checksummed = Address.toChecksummed.call(addr);
 * // "0x742d35Cc6634C0532925a3b844Bc9e7595f251e3"
 * ```
 */
export function toChecksummed(this: Address): Checksummed.Checksummed {
  return Checksummed.from(this);
}

File: src/primitives/Address/toChecksummed.go
package address

import (
	"encoding/hex"
	"strings"

	"golang.org/x/crypto/sha3"
)

func (a Address) ToChecksummed() string {
	lower := strings.ToLower(hex.EncodeToString(a[:]))

	hash := sha3.NewLegacyKeccak256()
	hash.Write([]byte(lower))
	hashBytes := hash.Sum(nil)
	hashHex := hex.EncodeToString(hashBytes)

	var result strings.Builder
	result.WriteString("0x")

	for i := 0; i < 40; i++ {
		ch := lower[i]
		if ch >= 'a' && ch <= 'f' {
			hv := hexCharToInt(hashHex[i])
			if hv >= 8 {
				result.WriteByte(ch - 32) // uppercase
			} else {
				result.WriteByte(ch)
			}
		} else {
			result.WriteByte(ch)
		}
	}

	return result.String()
}

func IsValidChecksum(s string) bool {
	if !IsValid(s) {
		return false
	}

	addr, err := FromHex(s)
	if err != nil {
		return false
	}

	checksummed := addr.ToChecksummed()
	if !strings.HasPrefix(s, "0x") {
		s = "0x" + s
	}

	return checksummed == s
}

func hexCharToInt(c byte) int {
	if c >= '0' && c <= '9' {
		return int(c - '0')
	}
	if c >= 'a' && c <= 'f' {
		return int(c - 'a' + 10)
	}
	if c >= 'A' && c <= 'F' {
		return int(c - 'A' + 10)
	}
	return 0
}
